File: From_Last_Year/Jutge/Week 7/Words_1.cpp
#include <iostream>
#include <string>
#include <vector>
#include <sstream>
#include <bits/stdc++.h>
using namespace std;

/*
WORD_1(n, A)
	let L be a list with an empty string inside
	
	for i = 1 to n do
		let L' be an empty list
		
		for all w in L do
			for all x in A do
				w' = x,w
				
				append w' to L'
		
		L = L'
	
	return L
*/

vector <string> word_1(int n, vector <char> A){
    
    vector <string> l = {""};
    vector <string> l2;

    sort(A.begin(), A.end());

    for(int i = 0; i < n; i++){
        l2.clear();
        
        for (auto &w : l) {
                
            for (auto &x : A){
                string new_w = (x+w);
                l2.push_back(new_w);
                }

            }
            
        l = l2;  

        }
        
    return l;
}

int main(){
    string s;
    getline(cin, s);

    int n = stoi(s);
    getline(cin, s);

    istringstream ss(s);

    vector <char> A;
    char x;

    while(ss >> x) A.push_back(x);

    for (auto &c : word_1(n, A)){
        cout << c << endl;
    }
     
        
        
    


    

}

File: From_Last_Year/Jutge/Week 7/Words_1.py
from jutge import read, read_line

"""
WORD_1(n, A)
	let L be an empty list
	insert an empty string to L
	
	sortA
	
	for i = 1 to n do
		let L' be an empty list
		
		for all w in L do
			for all x in A do
				w' = x,w
				
				append w' to L'
		
		L = L'
	
	return L
"""

n = read(int)
A = read_line().split()

def word_1(n, A):
	
	l = [""]
	A.sort()
	for i in range(0,n):
		l2 = []
		for w in l:
			for x in A:
				new_w = x+w
				l2.append(new_w)
		
		l = l2

	
	
	return l
	
	
for x in word_1(n, A):
	print(x)
	

	

